libdnf/query.cpp

```diff
@@ -0,0 +1,1656 @@
+    int cmp_type = f.get_cmp_type();
+
+    for (auto match_in : f.get_matches()) {
+        const char *match = match_in.str;
+        char *filter_vr = solv_dupjoin("0-", match, NULL);
+
+        for (Id id = 1; id < pool->nsolvables; ++id) {
+            if (!MAPTST(result, id))
+                continue;
+            char *e, *v, *r;
+            Solvable *s = pool_id2solvable(pool, id);
+            if (s->evr == ID_EMPTY)
+                continue;
+            const char *evr = pool_id2str(pool, s->evr);
+
+            pool_split_evr(pool, evr, &e, &v, &r);
+
+            if (cmp_type & HY_GLOB) {
+                if (fnmatch(match, r, 0))
+                    continue;
+                MAPSET(m, id);
+            }
+
+            char *vr = pool_tmpjoin(pool, "0-", r, NULL);
+
+            int cmp = pool_evrcmp_str(pool, vr, filter_vr, EVRCMP_COMPARE);
+
+            if ((cmp > 0 && cmp_type & HY_GT) ||
+                (cmp < 0 && cmp_type & HY_LT) ||
+                (cmp == 0 && cmp_type & HY_EQ)) {
+                MAPSET(m, id);
+            }
+        }
+        g_free(filter_vr);
+    }
+}
+
+void
+Query::Impl::filter_arch(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+    Id match_arch_id = 0;
+    for (auto match_in : f.get_matches()) {
+        const char *match = match_in.str;
+        if (f.get_cmp_type() & HY_EQ) {
+            match_arch_id = pool_str2id(pool, match, 0);
+            if (match_arch_id == 0)
+                continue;
+        }
+
+        for (Id id = 1; id < pool->nsolvables; ++id) {
+            if (!MAPTST(result, id))
+                continue;
+            Solvable *s = pool_id2solvable(pool, id);
+            if (f.get_cmp_type() & HY_EQ) {
+                if (match_arch_id == s->arch)
+                    MAPSET(m, id);
+                continue;
+            }
+            const char *arch = pool_id2str(pool, s->arch);
+            if (f.get_cmp_type() & HY_GLOB) {
+                if (fnmatch(match, arch, 0) == 0)
+                    MAPSET(m, id);
+                continue;
+            }
+        }
+    }
+}
+
+void
+Query::Impl::filter_sourcerpm(const Filter & f, Map *m)
+{
+    Pool *pool = dnf_sack_get_pool(sack);
+
+    for (auto match_in : f.get_matches()) {
+        const char *match = match_in.str;
+
+        for (Id id = 1; id < pool->nsolvables; ++id) {
+            if (!MAPTST(result, id))
+                continue;
+            Solvable *s = pool_id2solvable(pool, id);
+
+            const char *name = solvable_lookup_str(s, SOLVABLE_SOURCENAME);
+            if (name == NULL)
+                name = pool_id2str(pool, s->name);
+            if (!g_str_has_prefix(match, name)) // early check
+                continue;
+
+            DnfPackage *pkg = dnf_package_new(sack, id);
+            const char *srcrpm = dnf_package_get_sourcerpm(pkg);
+            if (srcrpm && !strcmp(match, srcrpm))
+                MAPSET(m, id);
+            g_object_unref(pkg);
```

Comment:
Ditto here, `g_autoptr(DnfPackage) pkg` ?

Reply:
Do we have `g_autoptr` in EL 7?

Reply:
Yes.  `git grep g_autofree` in the current source.

For ostree though, since Debian stable doesn't have it, we use the backports in <https://github.com/GNOME/libglnx/>.  That's not relevant here though fortunately.